packages/runtime/container-runtime/src/test/gc/gcTelemetryTracker.spec.ts

```diff
@@ -0,0 +1,737 @@
+	mixinMonitoringContext,
+	MonitoringContext,
+	ChildLogger,
+} from "@fluidframework/telemetry-utils";
+import {
+	GCNodeType,
+	GCTelemetryTracker,
+	defaultSessionExpiryDurationMs,
+	oneDayMs,
+	disableSweepLogKey,
+	UnreferencedStateTracker,
+	cloneGCData,
+} from "../../gc";
+import { pkgVersion } from "../../packageVersion";
+import { BlobManager } from "../../blobManager";
+
+export const configProvider = (settings: Record<string, ConfigTypes>): IConfigProviderBase => ({
+	getRawConfig: (name: string): ConfigTypes => settings[name],
+});
+
+describe("GC Telemetry Tracker", () => {
+	const defaultSnapshotCacheExpiryMs = 5 * 24 * 60 * 60 * 1000;
+	const sweepTimeoutMs = defaultSessionExpiryDurationMs + defaultSnapshotCacheExpiryMs + oneDayMs;
+	const inactiveTimeoutMs = 500;
+
+	// Nodes in the reference graph.
+	const nodes: string[] = ["/node1", "/node2", "/node3", "/node4"];
+
+	const testPkgPath = ["testPkg"];
+	// The package data is tagged in the telemetry event.
+	const eventPkg = { value: testPkgPath.join("/"), tag: TelemetryDataTag.CodeArtifact };
+
+	let injectedSettings: Record<string, ConfigTypes> = {};
+	let mockLogger: MockLogger;
+	let mc: MonitoringContext;
+	let clock: SinonFakeTimers;
+	let unreferencedNodesState: Map<string, UnreferencedStateTracker> = new Map();
+	let telemetryTracker: GCTelemetryTracker;
+
+	function createTelemetryTracker(
+		enableSweep: boolean,
+		isSummarizerClient = true,
+	): GCTelemetryTracker {
+		// Node types are as follows based on the path:
+		// Path starting with "/_blobs" - blob.
+		// Path with one part such as "/id1" - data stores.
+		// Path with two parts such as "/id1/id2" - sub data stores.
+		// Everything else - other.
+		const getNodeType = (nodePath: string) => {
+			if (nodePath.split("/")[1] === BlobManager.basePath) {
+				return GCNodeType.Blob;
+			}
+			if (nodePath.split("/").length === 2) {
+				return GCNodeType.DataStore;
+			}
+			if (nodePath.split("/").length === 3) {
+				return GCNodeType.SubDataStore;
+			}
+			return GCNodeType.Other;
+		};
+		const tracker = new GCTelemetryTracker(
+			mc,
+			{ inactiveTimeoutMs, sweepTimeoutMs: enableSweep ? sweepTimeoutMs : undefined },
+			isSummarizerClient,
+			false /* gcTombstoneEnforcementAllowed */,
+			{ createContainerRuntimeVersion: pkgVersion },
+			getNodeType,
+			(nodeId: string) => unreferencedNodesState.get(nodeId),
+			async (nodeId: string) => testPkgPath,
+		);
+		return tracker;
+	}
+
+	/**
+	 * For each node in nodeIds, add an entry in `unreferencedNodesState` indicating that the node was
+	 * just unreferenced.
+	 */
+	function markNodesUnreferenced(nodeIds: string[]) {
+		nodeIds.forEach((nodeId) => {
+			unreferencedNodesState.set(
+				nodeId,
+				new UnreferencedStateTracker(
+					Date.now(),
+					inactiveTimeoutMs,
+					Date.now(),
+					sweepTimeoutMs,
+				),
+			);
+		});
+	}
+
+	// Mock node loaded and changed activity for the given nodes.
+	function updateNodes(nodeIds: string[]) {
```

Comment:
nit: If this is a mock action taken during test setup/simulation I'd call it something more clear like `mockNodeChanges` or something